Apply a-priori algorithm to find frequently occurring items from given data   and   generate   strong   association   rules   using   support   and confidence thresholds.For Example: Market Basket Analysis

29 Aug 2020

In [1]:
import numpy as np
import pandas as pd
from apyori import apriori

In [2]:
data=pd.read_csv("mba.csv",header=None)
print(data.head())

              0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                  5     6               7             8             9   \
0  whole wheat flour  yams  cottage cheese  energy drink  tomato juice   
1                NaN   NaN             NaN           NaN           NaN   
2                NaN   NaN             NaN           NaN           NaN   
3                NaN   NaN             NaN           NaN           NaN   
4                NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey

In [3]:
obs=[]
for i in range(len(data)):
    obs.append([])
    for j in range(20):
        obs[-1].append(str(data.values[i,j]))
print(obs[:2])

[['shrimp', 'almonds', 'avocado', 'vegetables mix', 'green grapes', 'whole wheat flour', 'yams', 'cottage cheese', 'energy drink', 'tomato juice', 'low fat yogurt', 'green tea', 'honey', 'salad', 'mineral water', 'salmon', 'antioxydant juice', 'frozen smoothie', 'spinach', 'olive oil'], ['burgers', 'meatballs', 'eggs', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan', 'nan']]



- Support(I)=(No. of txn with I)/(total txn)

- Confidence(I1->I2)=(No. of txn with I1 & I2)/(No. of txn with I1)

- Lift(I1->I2)=Confidence(I1->I2)/Support(I2)


In [4]:
assoc=apriori(obs,min_support=0.005,min_confidence=0.3,min_lift=2,min_length=2)
assoc=list(assoc)

In [5]:
print(len(assoc))

167


In [6]:
print(assoc[1])

RelationRecord(items=frozenset({'mushroom cream sauce', 'escalope'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)])


RelationRecord(items=frozenset({'mushroom cream sauce', 'escalope'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)])

According to the above association:
- 0.5% txn contain M and E. (S)
- 30% of the txn containing M also contains E. (C)
- E is 3.79 times more likely to bought by customers who bought M compared to the default sale of E. 